### DISCUSSION TITLE
How to understand the parameter correction code in ReverseEnginerring module?

### DISCUSSION THREAD

**Discussion by lavender-lee** [FreeCAD/FreeCAD#18670] [Code contribution discussions]
Considering the following code:

https://github.com/FreeCAD/FreeCAD/blob/cad2f33b4e02b92659de3fb27f686f91c94b447f/src/Mod/ReverseEngineering/App/ApproxSurface.cpp#L1004

According to my limited knowledge, the correction terms are given by:

![image](https://github.com/user-attachments/assets/d582a1db-c17c-49e5-902e-3177a25dee3e)

Using the Mathematica, we have

```mathematica
LinearSolve[
 {{Xu^2 + XP*Xuu, Xu*Xv + XP*Xuv},
  {Xu*Xv + XP*Xuv, Xv^2 + XP*Xvv}}, {XP*Xu, XP*Xv}]
```
![image](https://github.com/user-attachments/assets/2f1d2683-110c-4da4-99ec-16a66dbedbd6)

where  $XP = X - P$

This is different with  the implement of code. Why? 

@wwmayer Could you give me an explaination, thanks.

**Comment by wwmayer** [comment]
I don't have access to the original literature any more. But the paper the formula was taken from assumed/demanded that the expression Xu * Xv + (X - P) * Xuv is equal to 0.

With this in mind the matrix can be simplified to:
[(Xu * Xu + (X - P) * Xuu, 0), (0, Xv * Xv + (X - P) * Xvv)]

The inverse of this matrix will be:
[(1/(Xu * Xu + (X - P) * Xuu), 0), (0, 1/(Xv * Xv + (X - P) * Xvv))]

Multiplied with the vector [(P - X) * Xu, (P - X) * Xv] this gives the expression as in the code.

**Reply by lavender-lee** [reply]
@wwmayer Got it, thanks a bunch :)

**Comment by wwmayer** [comment]
Here you can find a short summary of the algorithm: https://www.mathematik.tu-darmstadt.de/media/mathematik/forschung/preprint/ps2pdf/1717.pdf
On page 11 / 12 you will find the terms for the parameter correction:
Suu = Xu * Xu + (X- P) * Xuu
Svv = Xv * Xv + (X - P) * Xvv
Suv = Xu * Xv + (X - P) * Xuv

Compared to Suu and Svv the value of Suv is quite small (but not equal to zero) and that's probably why it is ignored. This idea originates from Hoschek, 1988 (Optimal approximate conversion of spline surfaces.)

The idea of the parameter correction is to find a better matching surface point to a given data point. The vector from this new surface point to the data point should be parallel to the surface normal at this point.

With this code you can verify that the parameter correction works as expected:
```
            pclBSplineSurf->D2(fU, fV, PntX, Xu, Xv, Xuu, Xvv, Xuv);
            {
                gp_Vec X(PntX.X(), PntX.Y(), PntX.Z());
                gp_Vec n1 = Xu ^ Xv;
                gp_Vec n2 = P - X;
                if (n1.Dot(n2) < 0) {
                    n2.Reverse();
                }
                double angle = Base::toDegrees(n1.Angle(n2));
                Base::Console().Message("Angle: %f \n", angle);
            }
```

With the parameter correction as it is currently implemented the results are already quite good (except of some cases close to the boundary of the surface).

However, when using the correct formula without ignoring Suv the results are always better, i.e. the angle between the two vectors is always smaller.

The code for the improved parameter correction is:
```
            double Suu = Xu * Xu + (X - P) * Xuu;
            double Svv = Xv * Xv + (X - P) * Xvv;
            double Suv = Xu * Xv + (X - P) * Xuv;
            double factor = 1.0 / (Suu * Svv - Suv * Suv);

            fDeltaU = factor * (Svv * (X - P) * Xu - Suv * (X - P) * Xv);
            fDeltaV = factor * (Suu * (X - P) * Xv - Suv * (X - P) * Xu);
```

**Reply by lavender-lee** [reply]
Thank you very very much for your explanation and paper. @wwmayer
